otri/analysis/convergence.py

```diff
@@ -0,0 +1,221 @@
+            # Remove atoms
+            self.__atoms[0] = self.__atoms[1] = None
+
+        # Check if it passed the ratio interval
+        if th.str_to_datetime(data['datetime']) >= self.__next_interval:
+            # Reset ratio
+            self.__next_interval = th.str_to_datetime(data['datetime']) + self.__time_group
+            self.__ratio_sum = 0
+            self.__interval_counter += 1
+            self.__interval_atoms_counter = 0
+        self._push_data(data, index=index)
+
+    def _input_check_order(self) -> Sequence[int]:
+        '''
+        Defines the order for the inputs to be checked.
+        '''
+        return [0] if self.__counter % 2 == 0 else [1]
+
+
+def ratio_atom_func(avg_ratio: float, elements):
+    # Two elements: atom from stream 1 and atom from stream 2
+    new_atom = dict()
+    new_atom['close'] = (float(elements[0]['close'])/float(elements[1]['close']) / avg_ratio) - 1
+    return new_atom
+
+
+class ConvergenceAnalysis(Analysis):
+    '''
+    Calculates the ratio between two time series in different periods and returns its value and its variance.
+    '''
+
+    def __init__(self, group_resolution: timedelta = timedelta(hours=4), ratio_interval: timedelta = timedelta(days=1)):
+        '''
+        Parameters:\n
+            group_resolution : timedelta
+                Resolution to group atoms to. Must be greater than atoms' resolution.\n
+            ratio_interval : timedelta
+                Interval of time where to calculate the average ratio. Must be greater than group_resolution\n
+        '''
+        self.__group_resolution = group_resolution
+        self.__ratio_interval = ratio_interval
+
+    def execute(self, input_streams: Sequence[Stream]):
+        '''
+        Starts convercence analyis.\n
+
+        Parameters:\n
+            in_streams : Stream
+                Two time series streams to analyise, must contain same-interval atoms with 'close' key.
+        '''
+        # Prepare output_streams
+        output_streams = [Stream(), Stream()]
+        # Calculate ratios ever ratio_interval
+        convergence_net = FilterNet(layers=[
+            FilterLayer([
+                # Tuple extractor
+                # TODO: AVOID Tuple-extracting here
+                GenericFilter(
+                    inputs="s1",
+                    outputs="atoms1",
+                    operation=lambda element: element[1]
+                ),
+                GenericFilter(
+                    inputs="s2",
+                    outputs="atoms2",
+                    operation=lambda element: element[1]
+                )
+            ], EXEC_AND_PASS),
+            FilterLayer([
+                # To Lowercase
+                GenericFilter(
+                    inputs="atoms1",
+                    outputs="lower_s1",
+                    operation=lambda atom: kh.lower_all_keys_deep(atom)
```

Comment:
Codacy found an issue: [Lambda may not be necessary](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6092628)